File: src/main/java/tally/statement/Transaction.java
package tally.statement;

import tally.core.Account;
import tally.core.Balance;

import javax.annotation.Nullable;

public class Transaction {
  private final Balance balance;
  private final String description;
  private final Account account;
  private final Type type;
  @Nullable private Long balanceFromStart;
  @Nullable private Long balanceFromEnd;

  public enum Type {
    UNKNOWN,
    EXPENSE,
    INCOME,
    TRANSFER
  }

  public Transaction(Balance balance, String description, Account account, Type type) {
    this.balance = balance;
    this.description = description;
    this.account = account;
    this.balanceFromStart = null;
    this.balanceFromEnd = null;
    this.type = type;
  }

  public Balance getBalance() {
    return balance;
  }
  public String getDescription() {
    return description;
  }
  public Account getAccount() {
    return account;
  }

  @Nullable
  public Long getBalanceFromStart() {
    return balanceFromStart;
  }

  @Nullable
  public Long getBalanceFromEnd() {
    return balanceFromEnd;
  }

  public void setBalanceFromEnd(@Nullable Long balanceFromEnd) {
    this.balanceFromEnd = balanceFromEnd;
  }

  public void setBalanceFromStart(@Nullable Long balanceFromStart) {
    this.balanceFromStart = balanceFromStart;
  }

  public boolean isExpense() {
    return type == Type.EXPENSE;
  }

  public boolean isIncome() {
    return type == Type.INCOME;
  }

  @Override
  public String toString() {
    return String.format("Transaction [account=%s, balance=%s, description=%s,"
            +" balanceFromStart=%s, balanceFromEnd=%s]",
        account, balance, description, balanceFromStart, balanceFromEnd);
  }
}

File: server/kt/src/main/kotlin/com/cherba29/tally/statement/Transaction.kt
package com.cherba29.tally.statement

import com.cherba29.tally.core.Balance
import com.cherba29.tally.core.TreeNode

/**
 * For given account captures transaction entry within transaction statement.
 */
data class Transaction(
  val targetTreeNode: TreeNode,
  val balance: Balance,
  val description: String?,
  val type: Type,
  val balanceFromStart: Long?
)  : Comparable<Transaction> {
  /**
   * Order transactions by balance amount.
   */
  override fun compareTo(other: Transaction): Int {
    val eq: Int = balance.compareTo(other.balance)
    return if (eq != 0) eq
    else {
      if (targetTreeNode.name != other.targetTreeNode.name) {
        if (targetTreeNode.name < other.targetTreeNode.name) -1 else 1
      } else if (description != other.description) {
        if (description.orEmpty() < other.description.orEmpty()) -1 else 1
      } else if (type != other.type) {
        type.compareTo(other.type)
      } else {
        0
      }
    }
  }

  /**
   * Type of transaction to keep track of whether funds are incoming
   * or outgoing or just being reshuffled.
   */
  enum class Type {
    UNKNOWN,

    /**
     * Transfer between two internal accounts.
     */
    TRANSFER,

    /**
     * Transfer from external to internal account.
     */
    INCOME,

    /**
     * Transfer from internal to external account.
     */
    EXPENSE,
  }
}
